coffeescript-to-javascript-guidelines.md

```diff
@@ -0,0 +1,168 @@
+  hasProp = {}.hasOwnProperty;
+
+CoreProduct = (function(superClass) {
+  extend(CoreProduct, superClass);
+
+  function CoreProduct() {
+    return CoreProduct.__super__.constructor.apply(this, arguments);
+  }
+
+  CoreProduct.prototype.initialize = function() {
+    return CoreProduct.__super__.initialize.apply(this, arguments);
+  };
+
+  return CoreProduct;
+
+})(Product);
+
+// ---
+// generated by coffee-script 1.9.2
+```
+
+#### Refactored JavaScript
+We can change from CoffeeScript's classes to Backbone's classes, like so
+
+```javascript
+var CoreProduct = Product.extend({
+  initialize: function() {
+    // we also change from using from `CoreProduct.__super__` to `Product.prototype`
+    return Product.prototype.initialize.apply(this, arguments);
+  }
+});
+```
+
+### Variables
+#### CoffeeScript
+
+```coffee
+_ = require('underscore')
+Product = require('models/product')
+TerminalMixins = require('presenters/mixins/terminal')
+```
+
+#### Generated JavaScript
+This is what would be output from js2coffee
+
+```javascript
+var Product, TerminalMixins, _;
+
+_ = require('underscore');
+
+Product = require('models/product');
+
+TerminalMixins = require('presenters/mixins/terminal');
+
+// ---
+// generated by coffee-script 1.9.2
+```
+
+#### Refactored JavaScript
+We would move the variable declarations to the same line that they are initialised, like so
+
+```javascript
+var _ = require('underscore');
+var Product = require('models/product');
+var TerminalMixins = require('presenters/mixins/terminal');
+```
+
+### Existence checks
+#### CoffeeScript
+
+```coffee
+_hasFlightCode: ->
+  @get('flights')?.out?.flight?.code
+```
+
+#### Generated JavaScript
+This is what would be output from js2coffee
+
+```javascript
+_hasFlightCode: function() {
+  var ref, ref1, ref2;
+  return (ref = this.get('flights')) != null ? (ref1 = ref.out) != null ? (ref2 = ref1.flight) != null ? ref2.code : void 0 : void 0 : void 0;
+}
+
+// ---
+// generated by coffee-script 1.9.2
+```
+
+#### Refactored JavaScript
+This is fairly unreadable, we would remove the temporary `ref` variables and do the following
+
+```javascript
+_hasFlightCode: function() {
+  var flights = this.get('flights');
+  if(!(flights && flights.out && flights.out.flight && flights.out.flight.code)) {
+    return;
+  }
+  return flights.out.flight.code;
+}
+```
+
+### Binding
+#### CoffeeScript
+
+```coffee
+_.debounce((=> @trigger('scroll')), 100)
+```
+
+#### Generated JavaScript
+This is what would be output from js2coffee
+
+```javascript
+_.debounce(((function(_this) {
+  return function() {
+    return _this.trigger('scroll');
+  };
+})(this)), 100);
+```
+
+#### Refactored JavaScript
+Rather than wrapping & passing `this` into the scope of the callback, we would assign `this` to `self` outside the scope of `debounce`, like so
+
+```javascript
+var self = this;
+_.debounce(function() {
```

Comment:
The reason I'd favour bind would be for consistency across the app. It allows for anonymous or named functions.

```
_.debounce(
  _.bind(function(){
    return self.trigger('scroll');
  }, this)
, 10);
```

could easily be refactored to...

```
_.debounce( _.bind(_myScrollyFunction, this), 10);
```